### PR TITLE
IcingaDB cube: Add sort control

### LINKED ISSUE #101: Add sort control
Add sort control to sort host and service dimensions by value and state (severity).

### PR DESCRIPTION
**Pull request by sukhwinder33445**
resolves #101

### CODE REVIEW

**Review by lippserd — COMMENTED** [review]
I'm not really happy with this code as I can't see an algorithm, just a brute force approach to grouping and then sorting the results.

The ROLLUP result is a flat tree-like data structure, for which we already have special handling in rendering and now also in sorting. I think it would make sense from a code quality perspective to turn the result into a tree once and then work with it throughout the code.

Unoptimized code to build the tree:

```php
        $results = $this->cube->fetchAll();

        $perspective = end($this->dimensionOrder);
        $tree = [new TreeNode()];

        for ($i = 0; ; $i++) {
            $row = $results[$i];

            if ($row->$perspective !== null) {
                break;
            }

            $node = (new TreeNode())
                ->setValue($row);
            $parent = end($tree);
            $parent->appendChild($node);
            $tree[] = $node;
        }

        for ($i += 1; $i < count($results); $i++) {
            $row = $results[$i];

            $node = (new TreeNode())
                ->setValue($row);

            $anyNull = false;
            foreach ($this->dimensionOrder as $dimension) {
                if ($row->$dimension === null) {
                    $anyNull = true;
                    break;
                }
            }
            if ($anyNull) {
                array_pop($tree);
            }
            $parent = end($tree);
            $parent->appendChild($node);
            if ($anyNull) {
                $tree[] = $node;
            }
        }

        $tree = $tree[1];
        // Sort while yield possible from here on.
```

Easy sorting:

```php
$a = [
    'critical_unhandled' => 10,
    'critical' => 20,
    'warning_unhandled' => 5,
    'warning' => 10,
    /* ... */
];

$b = [
    'critical_unhandled' => 10,
    'critical' => 20,
    'warning_unhandled' => 5,
    'warning' => 20,
    /* ... */
];

$priorities = ['critical', 'warning', /* ... */];

function sorting(array $a, array $b, array $priorities): array
{
    $autoUnhandled = function () use ($priorities): Generator {
        foreach ($priorities as $priority) {
            yield $priority . '_unhandled';
            yield $priority;
        }
    };

    foreach ($autoUnhandled() as $column) {
        $comparison = $a[$column] <=> $b[$column];
        switch ($comparison) {
            case 1:
                return $a;
            case -1:
                return $b;
        }
    }

    // $a and $b are equal in terms of $priorities.
    return $a;
}

assert(sorting($a, $b, $priorities) === $b);
```

@nilmerg Do you also like to change this or review the code in its current state?

File: library/Cube/CubeRenderer.php [hunk 1]
```diff
@@ -185,13 +194,167 @@ public function render(View $view)
...
+                    $node = (new TreeNode())
+                        ->setValue($row);
+
+                    $anyNull = false;
+                    foreach ($this->dimensionOrder as $dimension) {
+                        if ($row->$dimension === null) {
+                            $anyNull = true;
+                            break;
+                        }
+                    }
```
Comment (nilmerg) [inline library/Cube/CubeRenderer.php:234]: empty line after this pls

File: library/Cube/CubeRenderer.php [hunk 2]
```diff
@@ -185,13 +194,167 @@ public function render(View $view)
...
+                    if ($anyNull) {
+                        array_pop($tree);
+                    }
```
Comment (nilmerg) [inline library/Cube/CubeRenderer.php:237]: empty line after this pls

File: library/Cube/CubeRenderer.php [hunk 3]
```diff
@@ -185,13 +194,167 @@ public function render(View $view)
...
+
+        foreach ($preparedResults as &$d2) {
+            if (! is_array($d2)) {
+                break;
+            }
+
+            $this->sort($d2[1]);
+        }
+    }
+
```
Comment (nilmerg) [inline library/Cube/CubeRenderer.php:357]: Please remove.

File: library/Cube/CubeRenderer/HostStatusCubeRenderer.php [hunk 4]
```diff
@@ -114,4 +115,52 @@ protected function getDetailsBaseUrl()
...
+            case ($a->hosts_unreachable - $a->hosts_unhandled_unreachable)
+                > ($b->hosts_unreachable - $b->hosts_unhandled_unreachable):
+                return 1;
+            case ($a->hosts_unreachable - $a->hosts_unhandled_unreachable)
+                < ($b->hosts_unreachable - $b->hosts_unhandled_unreachable):
+                return -1;
+            default:
+                return 0;
+        }
+    }
```
Comment (nilmerg) [inline library/Cube/CubeRenderer/HostStatusCubeRenderer.php:154]: Please remove.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Comment by nilmerg** [comment]
`hosts?dimensions=host.vars.location%2Chost.vars.app%2Chost.vars.department&showSettings=1&sort=severity%20desc` works.

`hosts?dimensions=host.vars.location%2Chost.vars.app%2Chost.vars.os&showSettings=1&sort=severity%20desc` does not:

```
Undefined array key 0

#0 /usr/share/icingaweb2-modules/cube/library/Cube/CubeRenderer.php(238): Icinga\Application\ApplicationBootstrap->Icinga\Application\{closure}(Integer, String, String, Integer)
#1 /usr/share/icingaweb2-modules/cube/library/Cube/CubeRenderer.php(210): Icinga\Module\Cube\CubeRenderer->sortBySeverity(Array)
#2 /usr/share/icingaweb2-modules/cube/library/Cube/Cube.php(339): Icinga\Module\Cube\CubeRenderer->render(Object(Icinga\Web\View))
#3 /usr/share/icingaweb2-modules/cube/application/views/scripts/cube-index.phtml(20): Icinga\Module\Cube\Cube->render(Object(Icinga\Web\View))
#4 /icingaweb2/library/Icinga/Web/View.php(235): include(String)
#5 /icingaweb2/library/vendor/Zend/View/Abstract.php(877): Icinga\Web\View->_run(String)
#6 /usr/share/icingaweb2-modules/cube/library/Cube/Web/Controller.php(178): Zend_View_Abstract->render(NULL)
#7 /usr/share/icingaweb2-modules/cube/application/controllers/HostsController.php(19): Icinga\Module\Cube\Web\Controller->renderCube()
#8 /icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Cube\Controllers\HostsController->indexAction()
#9 /icingaweb2/library/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch(String)
#10 /icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#11 /icingaweb2/library/Icinga/Application/Web.php(290): Zend_Controller_Front->dispatch(Object(Icinga\Web\Request), Object(Icinga\Web\Response))
#12 /icingaweb2/library/Icinga/Application/webrouter.php(105): Icinga\Application\Web->dispatch()
#13 /icingaweb2/public/index.php(4): require_once(String)
#14 {main}
```

File: library/Cube/CubeRenderer/HostStatusCubeRenderer.php [hunk 5]
```diff
@@ -114,4 +115,15 @@ protected function getDetailsBaseUrl()
     {
         return 'cube/hosts/details';
     }
+
+    protected function getSeveritySortColumns(): Generator
+    {
+        $columns = ['down', 'unreachable'];
+        foreach ($columns as $column) {
+            yield "hosts_unhandled_$column";
+        }
```
Comment (nilmerg) [inline library/Cube/CubeRenderer/HostStatusCubeRenderer.php:124]: Empty line afterwards pls

File: library/Cube/CubeRenderer/ServiceStatusCubeRenderer.php [hunk 6]
```diff
@@ -133,4 +134,15 @@ protected function getDetailsBaseUrl()
     {
         return 'cube/services/details';
     }
+
+    protected function getSeveritySortColumns(): Generator
+    {
+        $columns = ['critical', 'unknown', 'warning'];
+        foreach ($columns as $column) {
+            yield "services_unhandled_$column";
+        }
```
Comment (nilmerg) [inline library/Cube/CubeRenderer/ServiceStatusCubeRenderer.php:143]: Empty line afterwards pls

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]

**Pull request merged**